pkg/minishift/update/update.go

```diff
@@ -0,0 +1,305 @@
+*/
+
+package update
+
+import (
+	"bytes"
+	"crypto"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/google/go-github/github"
+	update "github.com/inconshreveable/go-update"
+	"github.com/minishift/minishift/pkg/util/archive"
+	pb "gopkg.in/cheggaaa/pb.v1"
+
+	githubutils "github.com/minishift/minishift/pkg/util/github"
+	"github.com/minishift/minishift/pkg/version"
+)
+
+const (
+	timeLayout  = time.RFC1123
+	githubOwner = "minishift"
+	githubRepo  = "minishift"
+)
+
+// CurrentVersion returns the current version of minishift binary installed on the system
+func CurrentVersion() (semver.Version, error) {
+	localVersion, err := version.GetSemverVersion()
+	if err != nil {
+		return semver.Version{}, err
+	}
+	return localVersion, nil
+}
+
+// LatestVersion returns the latest version of minishift binary available from upstream
+func LatestVersion() (semver.Version, error) {
+	latestVersion, err := getLatestVersionFromGitHub(githubOwner, githubRepo)
+	if err != nil {
+		return semver.Version{}, err
+	}
+	return latestVersion, nil
+}
+
+// IsNewerVersion compares the local and latest versions and returns a boolean
+func IsNewerVersion(localVersion, latestVersion semver.Version) bool {
+	if localVersion.Compare(latestVersion) < 0 {
+		return true
+	} else {
+		return false
+	}
+}
+
+// Update handles the update process by calling functions that download, verify, extract and replace the binary.
+// It returns an error if any of these functions fails at any point.
+func Update(output io.Writer, latestVersion semver.Version) error {
+	var extName string
+
+	if runtime.GOOS == "windows" {
+		extName = "zip"
+	} else {
+		extName = "tgz"
+	}
+
+	archiveName := fmt.Sprintf("minishift-%s-%s-%s.%s", latestVersion, runtime.GOOS, runtime.GOARCH, extName)
+	downloadLinkFormat := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/download/v%s/%s"
+
+	url := fmt.Sprintf(downloadLinkFormat, latestVersion, archiveName)
+	downloadedArchivePath, err := downloadAndVerifyArchive(url)
+	if err != nil {
+		return err
+	}
+
+	// Get the directory containing saved archive
+	dir := strings.Split(downloadedArchivePath, "minishift")[0]
+	defer os.RemoveAll(dir)
+
+	// Extract and replace the downloaded archive in place of current binary
+	binary, err := extractBinary(downloadedArchivePath)
+	if err != nil {
+		return err
+	}
+
+	err = updateBinary(binary)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("\nUpdated to minishift v%s successfully.\n", latestVersion)
+
+	return nil
+}
+
+// getLatestVersionFromGitHub gets the latest version of minishift available on GitHub.
+// It returns the version and error.
+func getLatestVersionFromGitHub(githubOwner, githubRepo string) (semver.Version, error) {
+	var (
+		release *github.RepositoryRelease
+		resp    *github.Response
+		err     error
+	)
+
+	client := githubutils.Client()
```

Comment:
So here we go, this will need to go via proxies as well, right? So internally it seems GitHub utils uses "httpClient = http.DefaultClient". So the aim is to make this call use proxies. We might beed to add the proxy flags to the update command, unless we assume that the env variables are set. For consistency with the start command we probably want to add the proxy flags.